Fix crash when serializing quotes of deleted posts for ActivityPub

ClearlyClaire opened the pull request:
Fixes #36362

This serializes quotes of deleted posts as `"quote": { "type": "Tombstone" }`, which is admittedly fairly weird, but is one of our only ways to convey the information as we do not store the URI of the quoted post.

This does not serialize the information to the compat attributes because I am not sure how other implementations handle that, and those typically expect a string, not an object.

ClearlyClaire committed:
Fix `"quote": { "type": "Tombstone" }` being ignored

ClearlyClaire committed:
Fix crash when serializing quotes of deleted posts for ActivityPub

Fixes #36362

asahilina commented:
I deployed this on dev.vt.social, in case you want to test against it: https://dev.vt.social/@admin/115326705855435832

asahilina commented:
I deployed this on vt.social, but the behavior is a bit wonky. When (I think) it federates correctly, I see this:

<img width="585" height="484" alt="Screenshot_20251008_140135" src="https://github.com/user-attachments/assets/26e57c34-2077-41af-a72d-0c32ca368ad7" />

But I was expecting "Post unavailable" instead.

There's also an issue where it looks like once a quote federates on its own, the deletion of the quoted post doesn't (and on top of that, the Rails/nginx caching means that even if this happens after the post was deleted, there's a time window where it keeps federating as existing to other instances):

<img width="587" height="611" alt="Screenshot_20251008_140337" src="https://github.com/user-attachments/assets/453abc58-4dae-4f2f-9070-feb62ef4e3eb" />

I'm guessing this is because I don't follow that account so there is no complete federation going on, but it feels a bit wonky.

ClearlyClaire committed:
Explicitly record Tombstone quotes as deleted

This adds a `deleted` state to the internal representation, but this does
not change the API, which already included such a state.

ClearlyClaire commented:
Good catch, pushed a commit to address that.

> There's also an issue where it looks like once a quote federates on its own, the deletion of the quoted post doesn't

Oh, good catch on this also. This is a separate issue, will address in a separate PR.

ClearlyClaire commented on app/lib/activitypub/activity/create.rb:
```diff
@@ -218,11 +218,11 @@ def process_tags
     approval_uri = @status_parser.quote_approval_uri
     approval_uri = nil if unsupported_uri_scheme?(approval_uri) || TagManager.instance.local_url?(approval_uri)
-    @quote = Quote.new(account: @account, approval_uri: approval_uri, legacy: @status_parser.legacy_quote?)
+    @quote = Quote.new(account: @account, approval_uri: approval_uri, legacy: @status_parser.legacy_quote?, state: @quote_uri.present? ? :pending : :deleted)
```
Yes, it only works for this reason. But then, I'm not sure what we should be doing if someone says they are quoting a `Tombstone` of something that isn't actually deleted.
